make more cases of mapreduce trimmable

JeffBezanson opened the pull request:
add test for trimming with limited dynamic dispatch

move mapreduce tests out of basic_jll test

mbauman commented:
There are two cases here.   I understand the "inner" mapreduce_empty cases best:

With the status quo — without these `where {F}`s — we are _already_ compiling everything meaningful.  We've either explicitly `@inline`d things or they are all small enough to inline into another function body where the `f`s are either called or are explicitly specialized like this.  There is, however one case that _wouldn't_ inline and is likely responsible for the trimming error here.  And that's [the error case](https://github.com/JuliaLang/julia/blob/55b1f3c41baeb6b0ccfc0ad89d7c941e6c8bc229/base/reduce.jl#L316-L321).  Perhaps we should _also_ (or _alternatively_) explicitly `@nospecialize f` there?

The "outer" `_mapreduce -> mapfoldl` and the like cases I don't understand.  Here surely we'd inline the `_mapreduce` body into the caller to just contain a call to a non-inlined `mapfoldl`.  At that level, we'd surely know that `f` is a constant, no?

JeffBezanson commented:
How about that. Indeed, for this specific case just the `mapreduce_empty_iter` method is sufficient.

JeffBezanson committed:
make more cases of mapreduce trimmable

add test for trimming with limited dynamic dispatch

move mapreduce tests out of basic_jll test

JeffBezanson committed:
Update test/trimming/hello.jl

Co-authored-by: Matt Bauman <mbauman@juliahub.com>

timholy commented on test/trimming/hello.jl:
```diff
@@ -3,8 +3,34 @@ const str = OncePerProcess{String}() do
+    b = all(x -> x >= 0, sorted_arr)
+    c = map(x -> x^2, sorted_arr)
+    # d = mapreduce(x -> x^2, +, sorted_arr) this doesn't work because of mapreduce_empty_iter having F specialized
+    # e = reduce(xor, rand(Int, 10))
```
I'm sure you noticed this, but
```julia
julia> Base.reduce_empty(xor, Int)
ERROR: ArgumentError: reducing over an empty collection is not allowed; consider supplying `init` to the reducer
Stacktrace:
 [1] _empty_reduce_error()
   @ Base ./reduce.jl:317
 [2] reduce_empty(f::Function, T::Type)
   @ Base ./reduce.jl:318
 [3] top-level scope
   @ REPL[1]:1
```

mbauman replied:
Sure, but we can compile errors just fine.  The part that makes it ironic is that all the concrete branches inline just fine.  But it's the error branch — where we actually _don't care_ what the `T` is (and could even `@nospecialize` it) — that breaks inlining and puts in a "dynamic" call that only goes one place.  The answer is to add a bunch of `::Type{T} where T`s:

```diff
diff --git a/base/reduce.jl b/base/reduce.jl
index bc22808e53..368291bc87 100644
--- a/base/reduce.jl
+++ b/base/reduce.jl
@@ -366,12 +366,12 @@ The value to be returned when calling [`mapreduce`](@ref), [`mapfoldl`](@ref) or
 [`mapfoldr`](@ref) with map `f` and reduction `op` over an empty array with element type
 of `T`. See [`Base.reduce_empty`](@ref) for more information.
 """
-mapreduce_empty(::typeof(identity), op, T) = reduce_empty(op, T)
-mapreduce_empty(::typeof(abs), op, T)      = abs(reduce_empty(op, T))
-mapreduce_empty(::typeof(abs2), op, T)     = abs2(reduce_empty(op, T))
+mapreduce_empty(::typeof(identity), op, ::Type{T}) where {T} = reduce_empty(op, T)
+mapreduce_empty(::typeof(abs), op, ::Type{T}) where {T}      = abs(reduce_empty(op, T))
+mapreduce_empty(::typeof(abs2), op, ::Type{T}) where {T}     = abs2(reduce_empty(op, T))

-mapreduce_empty(f::typeof(abs),  ::typeof(max), T) = abs(zero(T))
-mapreduce_empty(f::typeof(abs2), ::typeof(max), T) = abs2(zero(T))
+mapreduce_empty(f::typeof(abs),  ::typeof(max), ::Type{T}) where {T} = abs(zero(T))
+mapreduce_empty(f::typeof(abs2), ::typeof(max), ::Type{T}) where {T} = abs2(zero(T))
```
